### PR TITLE
[WIP] Research TCv4

### PR DESCRIPTION
**Pull request by ZephyrTFA**
Research Grid
----
- [x] Generation
- [x] Interactions
- [x] Research Point Usage
- [x] Completion Checking
- [x] Completion Handling

Research Nodes
----
- [x] Node Generation
- [x] Requirement Checking
- [x] Unlock Handling

This is planned to be a complete overhaul of the current research system which adds different types of researches, and a dedicated method of unlocking nodes. The old form of research was quite literally click button get gold, which is not only boring but provides little, if not negative, amounts of player interaction.

Changelog
---
todo

### CODE REVIEW

File: code/game/machinery/autolathe.dm [hunk 1]
```diff
@@ -110,12 +110,12 @@
...
 			blueprints += w
 	else
-		for(var/w in researched_designs)
-			var/datum/design/d = categorycheck ? SSresearch.techweb_design_by_id(w) : w
+		for(var/datum/design/design as anything in stored_research.unlocked_designs)
+			design = stored_research.unlocked_designs[design]
 			if(categorycheck)
-				if(!(selected_category in d.category))
+				if(!(selected_category in design.category))
 					continue
```
Comment (Fikou) [inline code/game/machinery/autolathe.dm]: stack these ifs maybe?

File: code/game/machinery/computer/Operating.dm [hunk 2]
```diff
@@ -43,24 +43,24 @@
...
-	for(var/i in linked_techweb.researched_designs)
-		var/datum/design/surgery/D = SSresearch.techweb_design_by_id(i)
-		if(!istype(D))
+	advanced_surgeries?.Cut()
+	for(var/datum/design/surgery/surgery as anything in research_web.unlocked_designs)
+		surgery = research_web.unlocked_designs[surgery]
+		if(!istype(surgery))
 			continue
-		advanced_surgeries |= D.surgery
+		advanced_surgeries.Add(surgery.surgery)
```
Comment (Fikou) [inline code/game/machinery/computer/Operating.dm]: this feels really hacky, can you elaborate why you did it like that?

File: code/game/mecha/mech_fabricator.dm [hunk 3]
```diff
@@ -219,11 +218,11 @@
...
 	if(set_name in part_sets)
-		for(var/v in stored_research.researched_designs)
-			var/datum/design/D = SSresearch.techweb_design_by_id(v)
-			if(D.build_type & MECHFAB)
-				if(set_name in D.category)
-					add_to_queue(D)
+		for(var/datum/design/design as anything in stored_research.unlocked_designs)
+			design = stored_research.unlocked_designs[design]
+			if(design.build_type & MECHFAB)
+				if(set_name in design.category)
```
Comment (Fikou) [inline code/game/mecha/mech_fabricator.dm]: stack these?

File: code/game/mecha/mech_fabricator.dm [hunk 4]
```diff
@@ -219,11 +218,11 @@
 
 /obj/machinery/mecha_part_fabricator/proc/add_part_set_to_queue(set_name)
 	if(set_name in part_sets)
```
Comment (Fikou) [inline code/game/mecha/mech_fabricator.dm:238]: early return

File: code/game/mecha/mech_fabricator.dm [hunk 5]
```diff
@@ -372,22 +360,22 @@
...
 	if(href_list["part"])
 		var/T = href_list["part"]
-		for(var/v in stored_research.researched_designs)
-			var/datum/design/D = SSresearch.techweb_design_by_id(v)
-			if(D.build_type & MECHFAB)
-				if(D.id == T)
+		for(var/datum/design/design as anything in stored_research.unlocked_designs)
+			design = stored_research.unlocked_designs[design]
+			if(design.build_type & MECHFAB)
+				if(design.id == T)
```
Comment (Fikou) [inline code/game/mecha/mech_fabricator.dm]: stack

File: code/game/mecha/mech_fabricator.dm [hunk 6]
```diff
@@ -372,22 +360,22 @@
...
 		var/T = href_list["add_to_queue"]
-		for(var/v in stored_research.researched_designs)
-			var/datum/design/D = SSresearch.techweb_design_by_id(v)
-			if(D.build_type & MECHFAB)
-				if(D.id == T)
-					add_to_queue(D)
+		for(var/datum/design/design as anything in stored_research.unlocked_designs)
+			design = stored_research.unlocked_designs[design]
+			if(design.build_type & MECHFAB)
+				if(design.id == T)
```
Comment (Fikou) [inline code/game/mecha/mech_fabricator.dm]: stack

File: code/game/mecha/mech_fabricator.dm [hunk 7]
```diff
@@ -413,14 +401,14 @@
...
 		var/T = href_list["part_desc"]
-		for(var/v in stored_research.researched_designs)
-			var/datum/design/D = SSresearch.techweb_design_by_id(v)
-			if(D.build_type & MECHFAB)
-				if(D.id == T)
-					var/obj/part = D.build_path
+		for(var/datum/design/design as anything in stored_research.unlocked_designs)
+			design = stored_research.unlocked_designs[design]
+			if(design.build_type & MECHFAB)
+				if(design.id == T)
```
Comment (Fikou) [inline code/game/mecha/mech_fabricator.dm]: stack

File: code/modules/ninja/suit/suit_attackby.dm [hunk 8]
```diff
@@ -1,58 +0,0 @@
-
-
-/obj/item/clothing/suit/space/space_ninja/attackby(obj/item/I, mob/U, params)
```
Comment (Fikou) [inline code/modules/ninja/suit/suit_attackby.dm:3]: why are we deleting this file? refilling ninja suit with radium/smoke powder/replacing power cell is nice

**Comment by Latentish** [comment]
Could we get some pictures or recordings of this new system? I'd appreciate being able to see it!

File: code/modules/research/bepis.dm [hunk 9]
```diff
@@ -143,12 +143,13 @@
 	flick("chamber_flash",src)
 	update_icon_state()
 	banked_cash = 0
-	if((gauss_real >= gauss_major) && (SSresearch.techweb_nodes_experimental.len > 0)) //Major Success.
-		say("Experiment concluded with major success. New technology node discovered on technology disc.")
-		new /obj/item/disk/tech_disk/major(dropturf,1)
-		if(SSresearch.techweb_nodes_experimental.len == 0)
-			say("Expended all available experimental technology nodes. Resorting to minor rewards.")
-		return
+	// TODO: BEPIS SHIT
```
Comment (Fikou) [inline code/modules/research/bepis.dm]: gwa gwa

File: code/modules/research/destructive_analyzer.dm [hunk 10]
```diff
@@ -112,48 +112,7 @@ Note: Must be placed within 3 tiles of the R&D Console
...
-		var/choice = input("Are you sure you want to destroy [loaded_item][user_mode_string]?") in list("Proceed", "Cancel")
-		if(choice == "Cancel")
-			return FALSE
-		if(QDELETED(loaded_item) || QDELETED(linked_console) || !user.Adjacent(linked_console) || QDELETED(src))
-			return FALSE
-		var/loaded_type = loaded_item.type
-		if(destroy_item(loaded_item))
-			linked_console.stored_research.add_point_list(point_value)
-			linked_console.stored_research.deconstructed_items[loaded_type] = point_value
+	//TODO: ITEMS INTO DEPARTMENTAL POINTS BASED ON COMPLEXITY AND VALUE
```
Comment (Fikou) [inline code/modules/research/destructive_analyzer.dm]: gwa gwa

File: code/modules/research/nanites/nanite_chamber.dm [hunk 11]
```diff
@@ -194,9 +194,9 @@
 /obj/machinery/nanite_chamber/attackby(obj/item/I, mob/user, params)
 	if(istype(I, /obj/item/multitool))
 		var/obj/item/multitool/multi = I
-		if(istype(multi.buffer, /obj/machinery/rnd/server))
-			var/obj/machinery/rnd/server/server = multi.buffer
-			linked_techweb = server.stored_research
+		if(istype(multi.buffer, /datum/research_web))
+			linked_techweb = multi.buffer
+			say("linked to server")
```
Comment (Fikou) [inline code/modules/research/nanites/nanite_chamber.dm]: use visible_message like the cloud controller

File: code/modules/research/research_disk.dm [hunk 12]
```diff
@@ -30,10 +30,6 @@
 	illustration = "bepis"
 	custom_materials = list(/datum/material/iron=300, /datum/material/glass=100)
 
-/obj/item/disk/tech_disk/major/Initialize()
```
Comment (Fikou) [inline code/modules/research/research_disk.dm:33]: gwa gwa

**Review by Fikou — CHANGES_REQUESTED** [review]
i'll review the new files later

File: code/game/machinery/computer/Operating.dm (hunk 2 above)
Comment (ZephyrTFA) [reply code/game/machinery/computer/Operating.dm]: It's the exact same thing, I just think .Add(X) looks nicer than |= X;
what OOP does to a man

File: code/modules/research/bepis.dm (hunk 9 above)
Comment (ZephyrTFA) [reply code/modules/research/bepis.dm]: this is still a draft my boy; WIP

**Comment by ZephyrTFA** [comment]
Agony

**Comment by ZephyrTFA** [comment]
> This pull request has conflicts, please resolve those before we can evaluate the pull request.

Yes I know it does, I'm only changing literally everything related to research

**Comment by ZephyrTFA** [comment]
The only thing stopping me from finishing this is Frontend bullshit. Why is TGUI garbage.

**Comment by ZephyrTFA** [comment]
maybe one day. But not today

**Comment by ZephyrTFA** [comment]
Not next week either

**Pull request closed without merging**